[SPARK-28169][SQL] Fix Partition table partition  PushDown failed by "OR" expression

AngersZhuuuu opened the pull request:
## What changes were proposed in this pull request?

Spark can't push down filter condition of **Or**:

Such as if I have a table default.test, his partition col is "dt",

If we use query : 

`select * from default.test `
` where dt=20190625 or (dt = 20190626 and id in (1,2,3) )`

In this case, Spark will resolve **Or** condition as one expression, and since this expr has reference of "id", then it can't been push down.

In my PR ,  for SQL like 
`select * from default.test`
 `where  dt = 20190626  or  (dt = 20190626  and xxx="")   `

For this **Or**  condition 
`or (dt = 20190626  or  (dt = 20190626  and xxx=""   )`

All expression about partition keys will be  extracted as an expression only contains partition expression  
 and  retain the original logical relationship of **And** & **Or**  like below :
`dt = 20190626  or  dt = 20190626`

Then this condition will Passed to HiveTableScanExec. Such predicate expressions can be pushed down  as expected .

For this PR, it will extract deep relation of OR expression and push down this condition 
## How was this patch tested?
Exist unit test

Commit:
[SPARK-28169] extract predicate expression deeply

AngersZhuuuu commented:
ok to test

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

wangyum reviewed (commented):
Thank you @AngersZhuuuu Is this issue has been fixed by [SPARK-27699](https://issues.apache.org/jira/browse/SPARK-27699)?

Commit:
resolve and deeply

AngersZhuuuu commented:
@wangyum I looked about #24598 , we are not same , what I want to do is to fix the problem of hive partition table's partition push down.  That pr is for ORC & Parquet filter condition push down.

AngersZhuuuu commented:
@dongjoon-hyun  @cloud-fan @HyukjinKwon

HyukjinKwon commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala:
```diff
@@ -237,21 +237,75 @@ private[hive] trait HiveStrategies {
+        }
+
+        val extractedPruningPredicates = extractPushDownPredicate(predicates, partitionKeyIds)
+          .filter(_ != null)
```
@AngersZhuuuu, just for clarification, this code path does support OR expression but you want to do a partial pushdown right? Considering it needs a lot of codes as @wangyum pointed out, I think we should better try to promote to use (or convert) Spark's Parquet or ORC. It looks like an overkill to me.

AngersZhuuuu replied:
@HyukjinKwon  What I do is to extract condition's about partition keys.For the old code :
` val (pruningPredicates, otherPredicates) = predicates.partition { predicate =>
          !predicate.references.isEmpty &&
          predicate.references.subsetOf(partitionKeyIds) }`

If in expression, there contains other key, it won't be a push to HiveTableScanExec, So what I to it to fix this situation, just extract all condition about partition keys, then push it to HiveTableScanExec, HiveTableScanExec will handle complex combine expressions.

AngersZhuuuu replied:
@HyukjinKwon 
Spark's Parquet or ORC is perfect,  and it can push down filter condition, but it can't resolve the problem  that when we read a Hive table,  our first behavior is scan, What this pr want to do is to reduce the time of  resolve file info and partition metadata, and the file we scan.  Then the file num or partition num is big, it takes too long.

HyukjinKwon replied:
I think we convert Hive table reading operations into Spark's ones, via, for instance, `spark.sql.hive.convertMetastoreParquet` conf. If the diff is small, I might be fine but this does look like an overkill to me. I haven't taken a close look but it virtually looks like we need a fix like https://github.com/apache/spark/pull/24598

I won't object if some other committers are fine with that.

AngersZhuuuu replied:
@HyukjinKwon  I know that  it's better to us convert hive metastore to DataSource API, but it can't fix all situation. In our production env, we just  change hive data's default storage type  to orc.  For partition table, if different partition;s serde is not the same, Convert will failed thing using convert, it will check all partition by table level serde.

cloud-fan commented:
shall we just add an optimizer rule to do CNF conversion?

AngersZhuuuu commented:
> shall we just add an optimizer rule to do CNF conversion?

@cloud-fan I think it's not about optimizer, since this is special for partition push down, and origin rule in HiveStrategies is too simple.

cloud-fan commented:
CNF conversion can solve this problem, isn't it?

AngersZhuuuu commented:
> CNF conversion can solve this problem, isn't it?

Understand you, You mean In Optimizer level change condition  it to CNF , only this can't solve this pr's problem. And it may seriously affect other rules.

 In HiveStrategies to Change it to CNF can work well .

AngersZhuuuu commented:
> CNF conversion can solve this problem, isn't it?

PS:  First, I wish to convert predicate to CNF, but since I found that HiveTableScanExec can resolve complex partitionkeys's expression combine of **And** & **Or**  , so I just need to keep origin combine relation of partitionkey expression .

AngersZhuuuu commented:
@srowen Could you review this pr. And give some advise

srowen commented:
This part really isn't my area

AngersZhuuuu commented:
> This part really isn't my area

Hmm, sorry , I just recently pull request in  spark community, a lot is not clear.

cloud-fan commented:
This place is definitely not the only place that extracts partition predicates (e.g. `FileSourceStrategy`), I'm -1 to add a hack just in this place. I still prefer to add a CNF conversion rule to solve this problem for all the places, or other general solutions.

AngersZhuuuu commented:
@cloud-fan  
Since I am familiar with code about hive , I just looked about other strategies about  scan data. 
How about we convert filter expression set to CNF in  PhysicalOperation. Then change the way get partition expressions like below, since with CNF, just use **subSet** to judge is not enough:
    ` val partitionKeyFilters = `
        `  ExpressionSet(normalizedFilters`
           ` .filterNot(SubqueryExpression.hasSubquery(_))`
            `.filter(_.references.subsetOf(partitionSet))`

Commit:
fit scala style

Commit:
Fix problem of eslaped Or condition

Fix situation of  A is partition key
SELECT * FROM  A
WHERE A=1 OR B = 2

Int this case, we should ignore this condition

AngersZhuuuu commented:
@cloud-fan 
Checked a lot, if we make a rule for convert CNF, we still need to change some code in FileSourceStrategy and HiveStrategies or others.  
How about make it as a general method, and call it in each section where need to do this?

AngersZhuuuu commented:
@cloud-fan 
Checked a lot, if we make a rule for convert CNF, we still need to change some code in FileSourceStrategy and HiveStrategies or others.  
How about make it as a general method, and call it in each section where need to do this?

AngersZhuuuu commented:
@jerryshao 
Hi , could you help to review this , thanks.

Commit:
Add ExtractPartitionPredicates

Commit:
Fix scala stype

Commit:
fix scalastyle

Commit:
Fix scala style

AngersZhuuuu commented:
ok to test

Commit:
Change method style

Commit:
Add comment

Commit:
add ExtractPartitionPredicates to DataSourceStrategy

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -310,3 +310,97 @@ object PhysicalWindow {
+ *   FROM DEFAULT.PARTITION_TABLE
+ *   WHERE DT = 20190601 OR (DT = 20190602 AND C = "TEST")
+ *
+ * Where condition "DT = 20190601 OR (DT = 20190602 AND C = "TEST")"
```
Let's be more clear about the approach here. We try to weaken a predicate so that it only refers to partition columns, by removing some branches in AND. Let's also mention the corner case when there is no partition column attribute in the predicate(we should return Nil in this case?).

AngersZhuuuu replied:
@cloud-fan  
In this place it will return (DT = 20190601 OR DT = 20190602). 
But this whole condition will still return to filter.
What I want to do is purely to avoid  read unnecessary partitions. When this case we only read partition(dt=20190601 + dt=20190602), If we don't push down this, we will read all data.

In  condition " **(DT = 20190602 AND C = "TEST")** ", **DT = 20190602**  is **C = "TEST"**'s precondition. 

If the whole condition is **DT = 20190601 OR (DT = 20190602 OR C = "TEST")**. We should return null, since **DT = 20190602**  is not  **C = "TEST"**'s  constraint.

AngersZhuuuu replied:
@cloud-fan 
In my code. coming predicate Set[Expression] has a potential **AND** logical.
For one Expression, it will be   restricted by other same level Expression.
and  :

- if it is a combine of **AND** each side can be a constraint to others, so it one side is tenable, it can return a tenable condition.
- if it is a combine of **OR**, if one side is out of control(such as have no condition about partition cols) this whole **OR** Expression should return NONE.  Only  when both side of **OR** 's child is reasonable, it can return a  tenable combine of **OR**. 
- if it 's a multilayer nested Expression combined by BinaryOperator.  It will visit the lowest level, if it found one level's **OR** Expression is untenable, it will break this Expression totally and return null.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/planning/patterns.scala:
```diff
@@ -310,3 +310,97 @@ object PhysicalWindow {
+ */
+object ExtractPartitionPredicates extends Logging {
+
+  private type ReturnType = Seq[Expression]
```
We don't need this type alias as `Seq[Expression]` is short.

AngersZhuuuu replied:
I will change it. Thanks

AngersZhuuuu replied:
Done, Thanks.

Commit:
remove return type since it's short

AngersZhuuuu commented:
> This place is definitely not the only place that extracts partition predicates (e.g. `FileSourceStrategy`), I'm -1 to add a hack just in this place. I still prefer to add a CNF conversion rule to solve this problem for all the places, or other general solutions.

Checked previous activity for the CNF normalization in below issue:
https://issues.apache.org/jira/browse/SPARK-6624

Seems we can't make it as an Optimizer Rule,  it should only work for some special point such as partition predicate and filter push down。

924060929 commented:
Does this PR no process? 
Why no merge my PR https://github.com/apache/spark/pull/25688

AmplabJenkins commented:
Can one of the admins verify this patch?
